### ISSUE TITLE
BarbarianBonus UI Problem

### ISSUE THREAD

**Issue by skodkim** [Thalassicus/cep-bnw#71]
(formerly: [3.10.1] BarbarianBonus not working?)

Hi

For the last coulple of releases I've had a suspission that there was something wrong with the BarbarianBonus set in CEP_Options.sql so today I started testing using only a clean install (no other mods in directory, cache and moduserdata cleared) of CEP/CAT/CIN 3.10.1. I started a new game and placed a barbarian unit using the Fire Tuner to test.

With default values in CEP_Options.sql I get no bonus at all (based on Visuals). Look at the attached image.

I also tried setting BarbarianBonus to 150 for all difficulty levels. The result is the same: No bonus based on the in game Visuals.

\Skodkim

![civ5screen0036](https://f.cloud.github.com/assets/6018459/1755840/0bfcb252-666b-11e3-9c72-cc918a053454.jpg)

**Comment by GrantSP** [comment]
I have a different experience. I usually play at King level so the values here are for King.

In vanilla the BarbarianBonus is 30 and it correctly shows as 30 in the tooltip.
In CEP it is 15 and the database reflects that change yet the tooltip shows it as 30

The lua code in the core files looks like it should pull the value from that table yet appears not to.
See EnemyUnitPanel.lua lines 1223 - 1234

I have noticed in vanilla games the tooltip changing to reflect the different setting, 70 for Settler etc. Also on Deity, with no bonus, the tooltip shows nothing instead of Bonus VS Barbarians 0% so maybe that is something to look at.

**Comment by skodkim** [comment]
So you're saying its "just" a display issue?

\Skodkim

**Comment by GrantSP** [comment]
I wouldn't be so bold as to make such an assertion. I'm just saying
something fishy is happening. It is more than just the bonus not working.

**Comment by GrantSP** [comment]
Maybe you and I should do a couple of more tests. Try placing a barbarian on a tile and seeing whaat the tooltip shows and then how long it actually takes to defeat it. Then place a unit from one of the CS in the game on the same tile and see if there is any difference.
That way the same game values are in play and the only difference should be whether it is a barbarian or a minorciv.
I guess we could also test against a majorciv, but they probably have other bonuses on them as well.
Just a thought.

**Comment by skodkim** [comment]
Off to Work now but I'll see if I can find time later.

I'd propose another setup though. Start a game with default CEP settings and place # barbarian warriors and defeat them with warriors of your own. Do it on grasslands so they don't have other bonusses (maximizes the difference between our bonus and theirs). 

Start another game doing the same but this time with barbarianbonus set to 100 or some other "high" value.

Compare the results. Should be fairly quick using the fire tuner.

\Skodkim

**Comment by GrantSP** [comment]
Ok, done a few tests and these are the conclusions I have come to.
The BarbarianBonus DOES do exactly as it is supposed to.
The values displayed in the tooltip are an accurate reflection of the strengths of each unit, "Approx DMG inflicted" & the Green/Red vertical bars seem to show the correct values.
The text underneath "Bonus VS Barbarians X%" seems to be using a hard-coded value from somewhere, not sure where. Or the percentage is being modified somehow for display and comes out differently. I think if you play on Deity it doesn't show that text.

I still don't know why these, and other, values are duplicated in CEAI_FixTables.lua. Beyond my skill level I'm afraid.

**Comment by skodkim** [comment]
That was fast :-)

Good that its only a UI ting. UI is important but less so than gameplay.

\Skodkim

**Comment by Thalassicus** [comment]
> I still don't know why these, and other, values are duplicated in CEAI_FixTables.lua. Beyond my skill level I'm afraid.

BNW introduced a bug in the load order:
1. most sql data loads
2. overridden interface files check the data
3. "Start Game" data loads

Overridden interface files replace things like the unit combat panel. The third part includes anything on the "Start Game" screen, mainly HandicapInfos, Worlds, and GameSpeeds. This means lua files get the wrong information if we mod those data tables. I fixed this by manually creating lua tables identical to the sql tables.

I reported this load-order bug in the summer, so the problem _might_ have been fixed in the fall patch. I haven't tested it recently.

**Comment by skodkim** [comment]
@Thalassicus 

Thanks for the information but I don't really understand if you can fix this?

\Skodkim

**Comment by Thalassicus** [comment]
I don't know. The unit combat panel is very buggy because of its heavily hardcoded design. The mod does not currently override the unit combat panel, so any bugs with it exist in the unmodded game.

I could probably fix this once we add an override to that panel back in from Civup pre-BNW. However, I'm focusing on gameplay more than visuals this past half year. I haven't had time to do much interface stuff since BNW's release. The interface is very difficult and frustrating to modify.

**Comment by GrantSP** [comment]
I think we can probably close this. As it seems it is not really an 'issue', more of a desired UI change.

**Comment by stackpoint** [comment]
As I was chasing another issue I discovered this bug again and I found the section of the UI code that deals with the barbarian issues. EnemyUnitPanel.lua lines 1225 & 1964:

``` lua
iModifier = GameInfo.HandicapInfos[Game:GetHandicapType()].BarbarianBonus;
```

the only changes that needs to be made to this file is:

``` lua
iModifier = CepGameInfo.HandicapInfos[Game:GetHandicapType()].BarbarianBonus;
```

which is why we have the separate FixTables.lua file for the handicaps file.

**Comment by GrantSP** [comment]
@stackpoint are you on to something?

**Comment by stackpoint** [comment]
Thal basically discussed the bug in detail [here](http://forums.civfanatics.com/showpost.php?p=12674277&postcount=9) and I basically worked it from there. The only issue we have to wait on is whether or not we want to implement GEM's version of EnemyUnitPanel.lua or not.

**Comment by GrantSP** [comment]
I tell you what, that man is a damn fine coder!

**Comment by stackpoint** [comment]
Fixed here: https://github.com/Thalassicus/cep-bnw/commit/672b3bb2e758ea4ae2ddf1349439203d67e95f58
